controllers/k8ssandra/medusa_reconciler.go

```diff
@@ -346,3 +347,34 @@ func (r *K8ssandraClusterReconciler) getOperatorNamespace() string {
 	}
 	return operatorNamespace
 }
+
+// setupMedusaCleanup adds owner references to ensure that the remote resources created by reconcileMedusa are correctly
+// cleaned up when the CassandraDatacenter is deleted. We do that in a second pass because the CassandraDatacenter did
+// not exist yet at the time those resources were created.
+func (r *K8ssandraClusterReconciler) setupMedusaCleanup(
+	ctx context.Context,
+	kc *k8ssandraapi.K8ssandraCluster,
+	dc *cassdcapi.CassandraDatacenter,
+	remoteClient client.Client,
+	logger logr.Logger,
+) result.ReconcileResult {
+	// Note: this ConfigMap is an edge case because it is not DC-specific. If two CassandraDatacenters are in the same
+	// namespace, they would share the same ConfigMap, so setting one of them as the owner is wrong.
+	// However, this is an unlikely scenario (DCs are usually in different contexts, let alone namespaces). Also, if we
+	// delete one DC and the other still needs the ConfigMap, it will be recreated.
+	configMapKey := client.ObjectKey{
+		Namespace: dc.Namespace,
+		// see pgk/medusa/reconcile.go
+		Name: fmt.Sprintf("%s-medusa", kc.SanitizedName()),
+	}
+	result := setDcOwnership(ctx, dc, configMapKey, &corev1.ConfigMap{}, remoteClient, r.Scheme, logger)
+	if result.Completed() {
```

Comment:
Issue: This doesn't seem like the right use of `Completed()`. Actually, I try not to use `Completed()` at all - it isn't clear what it means because the `callBackSoon` implementation of `ReconcileResult` also returns `Completed() == true`. 

I would use one of these:

```
	IsError() bool
	IsRequeue() bool
	IsDone() bool
```

If you have an `IsError()` or `IsRequeue()` then you should return out to bail on the reconcile loop and allow the controller to requeue it. Conversely, actually if it is `IsDone()` then you can continue the reconciliation. 

Right now, I'm pretty sure you'll end up bailing early on requeues which isn't what you want.

Reply:
> Conversely, actually if it is `IsDone()` then you can continue the reconciliation

No, I think what `IsDone()` means is "the current reconciliation loop is done", in other words: drop everything we're doing and exit the main reconciliation function without any requeues.

When we want to continue the reconciliation, we use `Continue()`,  and `Continue().Completed()` is false. In fact it's the only implementation that returns false, so `Completed()` means "_is not_ `Continue()`".

So I think the use of `Completed()` here is correct: we set the ownership of the ConfigMap;
- if that succeeded, it returns `Continue()` and we proceed with the ownership of the CronJob
- if that didn't succeed (it errored out, was requeued, or we decided the whole reconciliation loop was fulfilled), then we return without addressing the CronJob

Reply:
> The issue I have with Completed() is that it is unclear what is completed. It doesn't seem to match the use of Done(), so that's problematic IMO.

Done() and Completed() are different. Done() is indeed Completed also, but Error() or CallbackSoon() are not Done(), but are Completed(). 

Completed() means the current reconcile call is done as far as we can right now, but does not mean our entire reconciliation process is done. We might requeue through error or requeue request to process it again. Done() is exactly that, we will not return to that reconcile part without any external changes and this process will not requeue.

This process hasn't changed since 2020 from where it was copied from. The tests indicate that as well.

https://github.com/datastax/cass-operator/blob/v1.0.0/operator/internal/result/result_helper_test.go

Reply:
I don't think those tests are sufficient documentation to indicate that the reconcile.Result type is meant to work that way @burmanm. They are talking about "reconciling being complete" but that clearly isn't what `Completed()` is actually intended to mean. As you say, it is probably intended that "completed" here means that the current reconciliation run is completed, but reconciliation as a whole is not complete. 

Again, **what particular reconciliation action** is done or completed, or needs to be requeued is not clear from any of this logic, nor the tests. And for ReconcileObject, `Done()` is used to indicate that the action was successful. So that's why I suggest we make things more explicit by avoiding the use of `Completed()` and instead returning when something explicitly needs to be requeued or has errored.

Reply:
> what particular reconciliation action is done or completed [...] is not clear

As I understand it, the action under consideration is always the top-level reconciliation loop, for example `K8ssandraClusterReconciler.Reconcile()`. So if a component returns `Done()`, it should "bubble up" the stack: every caller stops what it was doing and immediately returns, until we reach `Reconcile()` and return `ctrl.Result{}` to the runtime.

@burmanm would you agree with that assessment?

Reply:
#1419